engine/src/lib.rs

```diff
@@ -201,14 +201,15 @@ mod contract {
     /// Deploy staged upgrade.
     #[no_mangle]
     pub extern "C" fn deploy_upgrade() {
-        let io = Runtime;
+        let mut io = Runtime;
         let state = state::get_state(&io).sdk_unwrap();
         require_owner_only(&state, &io.predecessor_account_id());
         let index = internal_get_upgrade_index();
         if io.block_height() <= index {
             sdk::panic_utf8(errors::ERR_NOT_ALLOWED_TOO_EARLY);
         }
         Runtime::self_deploy(&bytes_to_key(KeyPrefix::Config, CODE_KEY));
+        io.remove_storage(&bytes_to_key(KeyPrefix::Config, CODE_STAGE_KEY));
```

Comment:
I'm not sure that we should automatically remove staged code. It could cause an issue for us in the future if we ever have a non-trivial state migration. See [near-plugins documentation](https://github.com/aurora-is-near/near-plugins/blob/master/near-plugins/src/upgradable.rs#L117) for more information.

Reply:
Then we probably need an additional transaction for removing staged code manually?

Reply:
Couldn't we call `stage_upgrade` with empty input to clear the old staged code?

Reply:
> I'm not sure that we should automatically remove staged code. It could cause an issue for us in the future if we ever have a non-trivial state migration. See [near-plugins documentation](https://github.com/aurora-is-near/near-plugins/blob/master/near-plugins/src/upgradable.rs#L117) for more information.

Might be misunderstanding something, but doesn't` io.remove_storage(&bytes_to_key(KeyPrefix::Config, CODE_STAGE_KEY));`  just clear the upgrade index (ie. the block height at which the staged code is allowed to be deployed), and not the actual staged code?

Reply:
Ah, yes, my mistake! Thanks for pointing that out. This is a nice way to prevent applying an upgrade multiple times without removing the staged code!